### PR TITLE
Revert #8860 (removal of public constructors)

### PR DESCRIPTION
**Pull request by maxkatz6**
Binary compatibility is still a problem, as events changing, features adding, and we have to duplicate ctors every time to avoid breaking compatibility.
Also, we want to avoid events faking by developers, as it's not a scenario we can always support.

On the other hand, some of the events are supposed to be raised by the developers.
And for other events we don't have a nice replacement. 

This PR is 11.0 timeline compromise.

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028293-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by maxkatz6** [comment]
@AvaloniaUI/core alternatively, we could remove all of these ctors and make properties `init;` only. It will solve at least one problem with breaking changes.

File: src/Avalonia.Base/Input/DragEventArgs.cs [hunk 1]
```diff
@@ -32,7 +33,9 @@ public Point GetPosition(Visual relativeTo)
             return point;
         }
 
-        internal DragEventArgs(RoutedEvent<DragEventArgs> routedEvent, IDataObject data, Interactive target, Point targetLocation, KeyModifiers keyModifiers)
+        [Unstable]
+        [Obsolete("This constructor may be removed in 12.0. Consider replacing it with RawDragEvent.")]
```
Comment (grokys) [inline src/Avalonia.Base/Input/DragEventArgs.cs:37]: Aren't the raw events also considered unstable? I thought that the alternative was to use the headless platform for unit tests, and raise custom pointer events at all outside unit tests.

File: src/Avalonia.Base/Input/PointerDeltaEventArgs.cs [hunk 2]
```diff
@@ -7,7 +9,9 @@ public class PointerDeltaEventArgs : PointerEventArgs
     {
         public Vector Delta { get; set; }
 
-        internal PointerDeltaEventArgs(RoutedEvent routedEvent, object? source, 
+        [Unstable]
+        [Obsolete("This constructor may be removed in 12.0. Consider replacing it with RawPointerEventArgs or headless unit tests helpers.")]
```
Comment (grokys) [inline src/Avalonia.Base/Input/PointerDeltaEventArgs.cs:13]: What's the reason to mention the headless platform here but not on `DragEventArgs`?

**Review by grokys — COMMENTED** [review]
I think we should probably just link to an issue (or a page in the docs) which explains what the alternative is here?

AFAIK, these event args should only be used for unit testing, and unit testing should be done via the headless platform (though we need docs for that) so the issue should explain the situation and in the longer term link to the headless unit test docs?

File: src/Avalonia.Base/Input/PointerDeltaEventArgs.cs (hunk 2 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Input/PointerDeltaEventArgs.cs:13]: We don't have anything in the headless for the drag events iirc

Comment (maxkatz6) [reply src/Avalonia.Base/Input/PointerDeltaEventArgs.cs]: Added DragDrop method to the IHeadlessWindow, updated comment.

File: src/Avalonia.Base/Input/DragEventArgs.cs (hunk 1 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Input/DragEventArgs.cs]: Added DragDrop method to the IHeadlessWindow, updated obsolete message. Removed mentioning of raw events.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029054-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029064-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by grokys — APPROVED** [review]
LGTM, we still need some proper documentation on how to unit test with the headless platform but this should unblock the people who are unit testing without the headless platform for now.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029107-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by robloo** [comment]
@grokys 

> AFAIK, these event args should only be used for unit testing

Perhaps you are talking about a specific one. However, in general these constructors are needed more for derived and custom controls. That is the biggest roadblock -- not unit testing.

* https://github.com/AvaloniaUI/Avalonia/pull/8860#issuecomment-1312819342
 * https://github.com/AvaloniaUI/Avalonia/pull/9534#issuecomment-1328052233

**Comment by timunie** [comment]
I also faced this in my custom controls. I need to reimplement the same EventArgs to solve this, which is not great imo, but doable. I wonder if we can have an event factory, if the constructor should be kept internally.

**Comment by grokys** [comment]
Ok, we should probably put together a list of use-cases where creating pointer events is needed and try to come up with some solution that will enable these use-cases while also not requiring an API to create arbitrary pointer events.

**Comment by timunie** [comment]
I need the VectorEventargs for a control which looks like a gantt-chart. I need to do zooming or scrolling based on key states. Having VectorEventArgs available is what I used for now.

**Comment by grokys** [comment]
I don't think `VectorEventArgs` needs to have an internal constructor at all. The ones that might cause problems are the pointer event args, because various other parts of the framework rely on them to set state (pointerover, capture etc) and so raising arbitrary ones may cause problems (especially in future).

**Comment by timunie** [comment]
Understood.

**Pull request merged**